server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -541,19 +565,74 @@ private void createShard(ShardRouting shardRouting, ClusterState state) {
         assert shardRouting.initializing() : "only allow shard creation for initializing shard but was " + shardRouting;
 
         try {
+            final var shardId = shardRouting.shardId();
+
             final DiscoveryNode sourceNode;
             if (shardRouting.recoverySource().getType() == Type.PEER) {
                 sourceNode = findSourceNodeForPeerRecovery(state.routingTable(), state.nodes(), shardRouting);
                 if (sourceNode == null) {
-                    logger.trace("ignoring initializing shard {} - no source node can be found.", shardRouting.shardId());
+                    logger.trace("ignoring initializing shard {} - no source node can be found.", shardId);
                     return;
                 }
             } else {
                 sourceNode = null;
             }
+            final var primaryTerm = state.metadata().index(shardRouting.index()).primaryTerm(shardRouting.id());
+
+            final var pendingShardCreation = createOrRefreshPendingShardCreation(shardId, state.stateUUID());
+            createShardWhenLockAvailable(
+                shardRouting,
+                state,
+                sourceNode,
+                primaryTerm,
+                0,
+                0L,
+                ActionListener.runBefore(new ActionListener<>() {
+                    @Override
+                    public void onResponse(Boolean success) {
+                        if (Boolean.TRUE.equals(success)) {
+                            logger.debug("{} created shard with primary term [{}]", shardId, primaryTerm);
+                        } else {
+                            logger.debug("{} gave up while trying to create shard", shardId);
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        failAndRemoveShard(shardRouting, true, "failed to create shard", e, state);
+                    }
+                }, () -> {
+                    assert ThreadPool.assertCurrentThreadPool(ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME);
+                    pendingShardCreations.remove(shardId, pendingShardCreation);
+                })
+            );
+        } catch (Exception e) {
```

Comment:
Do we need to remove from `pendingShardCreations` here? I wonder if we should catch Exception in `createShardWhenLockAvailable` and let it call `listener.onFailure`? I prefer to have methods that either invoke a listener or throw, not both.

Reply:
Ah yes good catch. Changed to catching all exceptions within the method in d6188af.